c_src/filewatch_inotify.c

```diff
@@ -181,41 +182,57 @@ static void send_output(ErlDrvData self_)
 {
     struct instance *self = (struct instance *)self_;
 
+    size_t len = notif_set_get_length(self->notif_set);
+
     size_t pos = 0;
-    ErlDrvTermData buf[(MAX_EVENTS * EVENT_TUPLE_LENGTH) + BUFFER_TAIL_SIZE];
+    ErlDrvTermData buf[(len * EVENT_TUPLE_LENGTH) + BUFFER_TAIL_SIZE];
 
     buf[pos++] = ERL_DRV_PORT;
     buf[pos++] = driver_mk_port(self->port);
 
-    struct message *msg;
-    for(uint i = 0; i < self->n_events; ++i) {
-        msg = &(self->msgs[i]);
+    struct Message msg[len];
+    for(size_t i = 0; i < len; i++) {
+        msg[i].name = driver_alloc(NAME_MAX);
+    }
+    int num_entries_removed = 0;
+    struct Notif_set_itr *itr = notif_set_itr_init(self->notif_set);
+    while(notif_set_itr_next(itr, &msg[num_entries_removed])) {
         buf[pos++] = ERL_DRV_INT;
-        buf[pos++] = msg->wd;
+        buf[pos++] = msg[num_entries_removed].wd;
         buf[pos++] = ERL_DRV_STRING;
-        buf[pos++] = (ErlDrvTermData) msg->name;
-        buf[pos++] = strnlen(msg->name, NAME_MAX);
+        buf[pos++] = (ErlDrvTermData) msg[num_entries_removed].name;
+        buf[pos++] = strnlen(msg[num_entries_removed].name, NAME_MAX);
         buf[pos++] = ERL_DRV_TUPLE;
         buf[pos++] = BUFFER_TUPLE_SIZE;
+        
+        notif_set_itr_remove(itr);
+        num_entries_removed++;
     }
-
+    notif_set_itr_destroy(itr);
+ 
     /* Append values designating Erlang type */
     buf[pos++] = ERL_DRV_NIL;
     buf[pos++] = ERL_DRV_LIST;
-    buf[pos++] = self->n_events + 1; // +1 for nil
+    buf[pos++] = num_entries_removed + 1; // +1 for nil
     buf[pos++] = ERL_DRV_TUPLE;
     buf[pos++] = BUFFER_TUPLE_SIZE;
 
-    self->n_events = 0;
+    //Don't need to free any memory associated with notif_set 
+    //entries; they are freed automatically as they get removed. 
 
     erl_drv_output_term(driver_mk_port(self->port), buf, pos);
+
+    for(size_t i = 0; i < len; i++) {
+        driver_free(msg[i].name);
+    }
+
+    self->is_timer_set = false;
 }
 
 static void serialize_event(struct instance *self, const struct inotify_event *evt) {
-    struct message *msg = &(self->msgs[self->n_events++]);
-    msg->wd = evt->wd;
-    msg->len = strnlen(evt->name, NAME_MAX);
-    strncpy(msg->name, evt->name, NAME_MAX);
+    //Insert into notif_set to keep track of unique {wd, name} tuples.
+    notif_set_insert(self->notif_set,
+                            &(struct Message) {.wd = evt->wd, .name = (char *) evt->name});
```

Comment:
Since this is the only place that `notif_set_insert` is called I would just pass `wd` and `name` directly instead of wrapping them in a temporary object. As it stands, the first time I read this code I immediately had concerns over the ownership of this temporary object.